node/src/components/consensus/era_supervisor/era.rs

```diff
@@ -161,6 +163,75 @@ impl<I> Era<I> {
     pub(crate) fn set_paused(&mut self, paused: bool) {
         self.consensus.set_paused(paused);
     }
+
+    /// Computes rewards and returns the era report.
+    pub(crate) fn create_report(
+        &self,
+        terminal_block_data: TerminalBlockData<ClContext>,
+        timestamp: Timestamp,
+        protocol_config: &ProtocolConfig,
+    ) -> EraReport<PublicKey> {
+        let TerminalBlockData {
+            inactive_validators,
+        } = terminal_block_data;
+
+        // The total rewards are the ideal number of blocks (i.e. one per minimum-length round) in
+        // the actual duration of the era, times the BLOCK_REWARD constant, independent of how many
+        // blocks were actually created.
+        let era_duration = timestamp.saturating_diff(self.start_time);
+        let min_round_length = protocol_config.highway_config.min_round_length();
+        let number_of_min_length_rounds = era_duration / min_round_length;
+        let total_rewards: u64 = number_of_min_length_rounds
+            .checked_mul(BLOCK_REWARD)
+            .unwrap_or_else(|| {
+                // With BLOCK_REWARD set to one trillion and a round length of 1 minute, a single
+                // era would have to last 35 years before the total rewards would overflow a u64.
+                error!(%era_duration, %min_round_length, %BLOCK_REWARD, "total rewards overflow");
+                u64::MAX
+            });
+
+        if total_rewards == 0 {
+            return EraReport {
+                rewards: BTreeMap::new(),
+                equivocators: self.accusations(),
+                inactive_validators,
+            };
+        }
+
+        let is_active = |public_key: &PublicKey| {
+            !(self.accusations.contains(public_key)
+                || self.faulty.contains(public_key)
+                || inactive_validators.contains(public_key))
+        };
+
+        // The total rewards are distributed proportionally by weight. To make sure that the
+        // computation doesn't overflow we divide all weights by 2^64 if the numbers are too large.
+        let total_weight: U512 = self
+            .validators
+            .iter()
+            .filter_map(|(public_key, weight)| is_active(public_key).then(|| *weight))
+            .sum();
+        let shift = if total_weight.checked_mul(total_rewards.into()).is_none() {
+            64
+        } else {
+            0
+        };
+        let rewards = self
+            .validators
+            .iter()
+            .filter_map(|(public_key, weight)| {
+                is_active(public_key).then(|| {
+                    let reward = (weight >> shift) * total_rewards / (total_weight >> shift);
```

Comment:
Is it possible that `total_weight` can be 0? Probably not, because that would mean that nobody created any blocks in the era, and so it wouldn't end, right?

Reply:
IIRC, consensus would fail (`fatal!`) if it received set of validators without any weight but it's a good point and maybe we _should_ guard against such case here as well.

Reply:
`total_weight` here is not the weight of _all_ validators, though, just the active ones. Still, we'd expect at least one validator with nonzero weight to have been active if we're creating a switch block, I believe. Might be a good idea to add a comment to this effect, though, so it's clear that this has been considered.

Reply:
A validator without any weight couldn't have proposed a block.

Reply:
I agree, I'll add a check for that.

Reply:
Done in https://github.com/casper-network/casper-node/pull/1950/commits/6755fddf31a5df0fff7717c54b1317b76d84511e.